File: backend/internal/models/progress.go
package models

import "time"

type UserProgress struct {
	ID          int       `json:"id"`
	UserID      int       `json:"user_id"`
	LessonID    int       `json:"lesson_id"`
	Completed   bool      `json:"completed"`
	CompletedAt time.Time `json:"completed_at"`
	CreatedAt   time.Time `json:"created_at"`
}

type UpdateProgressRequest struct {
	LessonID  int  `json:"lesson_id" validate:"required"`
	Completed bool `json:"completed"`
}

type UserStats struct {
	CoursesCompleted int    `json:"courses_completed"`
	HoursLearned     int    `json:"hours_learned"`
	Rank             int    `json:"rank"`
	CurrentStreak    int    `json:"current_streak"`
	TotalXP          int    `json:"total_xp"`
}

type PathProgress struct {
	PathID          string  `json:"path_id"`
	TotalLessons    int     `json:"total_lessons"`
	CompletedLessons int    `json:"completed_lessons"`
	Percentage      float64 `json:"percentage"`
}

File: frontend/src/api/progress.ts
import client from './client'

export interface UserProgress {
    id: number
    user_id: number
    lesson_id: number
    completed: boolean
    completed_at: string
    created_at: string
}

export interface UserStats {
    courses_completed: number
    hours_learned: number
    rank: number
    current_streak: number
    total_xp: number
}

export interface PathProgress {
    path_id: string
    total_lessons: number
    completed_lessons: number
    percentage: number
}

export interface UpdateProgressRequest {
    lesson_id: number
    completed: boolean
}

export const progressApi = {
    updateProgress(data: UpdateProgressRequest) {
        return client.post<void>('/progress', data)
    },

    getStats() {
        return client.get<UserStats>('/progress/stats')
    },

    getPathProgress(pathId: string) {
        return client.get<UserProgress[]>(`/progress/paths/${pathId}`)
    }
}
